ui/panel_completion.py

```diff
@@ -0,0 +1,279 @@
+                    {"view_id": view_id, "completion_index": index},
+                )
+            ),
+            "<i> Mean Probability: {}</i>".format(completion["score"]),
+        ]
+
+    def open(self) -> None:
+        # TODO: show this side-by-side?
+        window = self.view.window()
+        if not window:
+            # error message
+            return
+
+        current_group = window.active_group()
+        if current_group == window.num_groups() - 1:
+            self._open_in_side_by_side(window)
+        else:
+            self._open_in_group(window, current_group + 1)
+
+    def update(self) -> None:
+        # TODO: show this side-by-side?
+        sheet = find_sheet_by_id(self.completion_manager.sheet_id)
+        if sheet is None:
+            return
+
+        mdpopups.update_html_sheet(
+            sheet=sheet,
+            name="Panel Completions",
+            contents=self.completion_content,
+            md=True,
+            css=self.CSS,
+            wrapper_class=self.CSS_CLASS_NAME,
+        )
+
+    def close(self) -> None:
+        # TODO: show this side-by-side?
+        window = self.view.window()
+        if not window:
+            return
+
+        sheet = find_sheet_by_id(self.completion_manager.sheet_id)
+        if sheet is None:
+            return
+
+        sheet.close()
+        if self.completion_manager.original_layout:
+            window.set_layout(self.completion_manager.original_layout)
+            erase_copilot_view_setting(self.view, "original_layout")
+
+    @staticmethod
+    def _prepare_popup_code_display_text(display_text: str) -> str:
+        # The returned completion is in the form of
+        #   - the first won't be indented
+        #   - the rest of lines will be indented basing on the indentation level of the current line
+        # The rest of lines don't visually look good if the current line is deeply indented.
+        # Hence we modify the rest of lines into always indented by one level if it's originally indented.
+        first_line, sep, rest = display_text.partition("\n")
+
+        if rest.startswith((" ", "\t")):
+            return first_line + sep + textwrap.indent(textwrap.dedent(rest), "\t")
+
+        return display_text
+
+    @staticmethod
+    def _synthesize(completions: Iterable[CopilotPayloadPanelSolution]) -> List[CopilotPayloadPanelSolution]:
+        return sorted(unique(completions, itemgetter("completionText")), key=itemgetter("score"), reverse=True)
+
+    def _open_in_group(self, window: sublime.Window, group_id: int) -> None:
+        window.focus_group(group_id)
+        sheet = mdpopups.new_html_sheet(
+            window=window,
+            name="Panel Completions",
+            contents=self.completion_content,
+            md=True,
+            css=self.CSS,
+            wrapper_class=self.CSS_CLASS_NAME,
+        )
+        self.completion_manager.sheet_id = sheet.id()
+
+    def _open_in_side_by_side(self, window: sublime.Window) -> None:
```

Comment:
When I suggested `side_by_side` I mean the ST4 side-by-side thing(when two views are opened together), this way we don't need t worry about restoring the layout.

Reply:
What you mean is a transient view. I don't think we can create one via existing API?

Reply:
Here is how it is handled in LSP https://github.com/sublimelsp/LSP/blob/5fc0db5640d1b6515ad4a6f58f6797613e012e06/plugin/locationpicker.py#L24  
Basically `sublime.ADD_TO_SELECTION | sublime.SEMI_TRANSIENT`

Reply:
That should be an idea solution I think. @TheSecEng

Reply:
This works for existing files and `window.open_file` this doesn't work for `window.new_file`

Reply:
Also the phantom approach is a horrible UX. It flickers every update. Either of you want to take a stab at it?

Reply:
> This works for existing files and `window.open_file` this doesn't work for `window.new_file`

Or altleazst I can't get it to wrk

Reply:
ok. let's create a dummy file for opening 🤣